airbyte-scheduler/persistence/src/main/java/io/airbyte/scheduler/persistence/job_tracker/TrackingMetadata.java

```diff
@@ -102,9 +110,38 @@
             metadata.put("volume_rows", syncSummary.getRecordsSynced());
           }
         }
+
+        final List<FailureReason> failureReasons = failureReasonsList(attempts);
+        if (!failureReasons.isEmpty()) {
+          metadata.put("failure_reasons", failureReasonsListAsJson(failureReasons).toString());
+          metadata.put("main_failure_reason", failureReasonAsJson(failureReasons.get(0)).toString());
+        }
       }
     }
     return metadata.build();
   }
 
+  private static List<FailureReason> failureReasonsList(final List<Attempt> attempts) {
+    return attempts
+        .stream()
+        .map(Attempt::getFailureSummary)
+        .flatMap(Optional::stream)
+        .map(AttemptFailureSummary::getFailures)
+        .flatMap(Collection::stream)
+        .toList();
+  }
+
+  private static ArrayNode failureReasonsListAsJson(final List<FailureReason> failureReasons) {
+    return Jsons.arrayNode().addAll(failureReasons
+        .stream()
+        .map(TrackingMetadata::failureReasonAsJson)
+        .toList());
+  }
+
+  private static JsonNode failureReasonAsJson(final FailureReason failureReason) {
+    final JsonNode jsonNode = Jsons.jsonNode(failureReason);
+    ((ObjectNode) jsonNode).remove("stacktrace");
```

Comment:
One thought that just occurred to me - instead of removing stacktrace and including everything else by default, we could alternatively explicitly choose which fields to include. 

I think it ends up mattering most if/when the FailureReason object changes - if we add a new field to it down the road, your solution will include it by default. This could be a good thing or a bad thing depending on what it is. 

If we're selective about which fields to include, a developer would need to explicitly add the field here for it, which could be an easy-to-overlook step. That said, it looks everywhere else in this file, we're pretty selective about which fields get included in the metadata that we send, so maybe it'd be best to take that approach here as well? It feels like a slightly more robust approach to me, because we wouldn't end up with issues if, say, somebody added a huge field to the FailureReason object without thinking about the job tracker and we suddenly started running into size limits (probably very unlikely but technically possible).

Thoughts?

Reply:
yeah i think this does make sense. We also probably don't even all of the existing fields even ignoring stacktrace, so it should be easier to add/remove fields if we list them explicitly here.